// @jacada-jbot/base/include/Model/LRepositoryElementPath.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var LRepositoryElementPath = /** @class */ (function () {
    function LRepositoryElementPath() {
        this.currentIndex = 0;
    }
    LRepositoryElementPath.prototype.add = function (repositoryHandler) {
        this._repositoryHandlers.push(repositoryHandler);
    };
    LRepositoryElementPath.prototype.getFirstHandler = function () {
        if (this._repositoryHandlers.length === 0)
            return null;
        return this._repositoryHandlers[0];
    };
    LRepositoryElementPath.prototype.next = function () {
        if (this.currentIndex + 1 >= this._repositoryHandlers.length)
            return null;
        return this._repositoryHandlers[++this.currentIndex];
    };
    LRepositoryElementPath.prototype.current = function () {
        return this._repositoryHandlers[this.currentIndex];
    };
    LRepositoryElementPath.prototype.isLast = function () {
        return (this.currentIndex + 1 >= this._repositoryHandlers.length);
    };
    LRepositoryElementPath.prototype.reset = function () { this.currentIndex = 0; };
    Object.defineProperty(LRepositoryElementPath.prototype, "repositoryHandlers", {
        get: function () { return this._repositoryHandlers; },
        enumerable: false,
        configurable: true
    });
    ;
    return LRepositoryElementPath;
}());
exports.default = LRepositoryElementPath;
