// @jacada-jbot/web/include/Model/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Constants = exports.BaseConstants = void 0;
var Constants_1 = require("@jacada-jbot/base/include/Model/Constants");
Object.defineProperty(exports, "BaseConstants", { enumerable: true, get: function () { return Constants_1.Constants; } });
// export BaseConstants;
exports.Constants = {
    /**
     * uniqueId represents a unique id for a coverageHandler and it is used for events.
     */
    JBOT_UNIQUE_ID_KEY: "uniqueId",
    /**
     * This attribute indicates that this element has a Coverage handler per the attribute value in applicationData.
     */
    JBOT_COVERAGE_HANDLER_ATTR: "jbot_coverage_handler",
    /**
     * this attr is to mark the elements that are registered on events on specific coverage
     */
    JBOT_REGISTERED_ATTR: "jbot_registered",
    /**
     * this attribute used to hold an guid that represent commination between main frame and frame via extension
     */
    ATTR_FRAME_GUID: "JBOT_FRAME_GUID",
    JBOT_ATTR_PREFIX: "jbot_",
    /**
     * This element can be highlighted
     */
    JBOT_HIGHLIGHT_ATTR_ID: "jbot_highlight_id",
    /**
     *  Attribute on the current highlighted element
    */
    ELEMENT_IS_CURRENTLY_HIGHLIGHTED: "jbot_currently_highlighted",
    EVENT_OVERLAY_CLICKED: "overlayClicked",
    IS_PREVENT_CLICK_ON_ELEMENT: "isPreventClickOnElement",
    EVENT_CAPTURE_MOUSE: "CaptureMouseEvent",
    PARENT_ID: "___parent___",
    LABELS_TOPIC: "Labels",
};
