// @jacada-jbot/web/include/CoverageHandlers/Web/ButtonWebControlCoverageHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GenericResult_1 = require("@jacada-jbot/base/include/Model/GenericResult");
const LogicalAction_1 = require("@jacada-jbot/base/include/Model/LogicalAction");
const WebCustomControlCoverageHandler_1 = require("./WebCustomControlCoverageHandler");
const EventDetails_1 = require("@jacada-jbot/base/include/Model/EventDetails");
/**
 * Coverage handler for a button
 */
class ButtonWebControlCoverageHandler extends WebCustomControlCoverageHandler_1.WebCustomControlCoverageHandler {
    constructor() {
        super(...arguments);
        this.type = "Button";
    }
    startRecording() {
        this.recordOverlayClickedEvent((event) => {
            let lAction = new LogicalAction_1.default(this.click.name, "");
            return lAction;
        }, (event) => {
            this.click();
        });
        return true;
    }
    subscribeToClicked() {
        this.listenToEvent("click", () => {
            return new EventDetails_1.default("clicked");
        });
    }
    /**
     * This method fires a click event on the jquery object
     * @returns a GenericResult struct
     */
    click() {
        this.node[0].click();
        return new GenericResult_1.GenericResult(true, null, null);
    }
    updateElementRepositoryNode(elementNode) {
        elementNode.name = this.getName() || elementNode.name;
        elementNode.id = this.node.attr("id") || this.getId() || elementNode.id;
    }
    getName() {
        const val = this.node.val();
        if (val) {
            const text = val.toString();
            return text;
        }
        //get text from inner <button> element
        const walker = document.createTreeWalker(this.node[0], NodeFilter.SHOW_TEXT);
        const nodesText = [];
        while (walker.nextNode()) {
            nodesText.push(walker.currentNode.textContent.trim());
        }
        // Join all texts with a space between them
        return nodesText.join(" ");
    }
    getId() {
        // Try to make the ID meaningfully by having the right prefix when id is auto generated
        // Try to see if we have text.
        const id = this.getName();
        if (id.length > 0) {
            return id;
        }
        return "";
    }
}
exports.default = ButtonWebControlCoverageHandler;
