// @jacada-jbot/web/include/CoverageHandlers/Web/FormWebControlCoverageHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const WebCustomControlCoverageHandler_1 = require("./WebCustomControlCoverageHandler");
/**
 * Coverage handler for a form
 */
class FormWebControlCoverageHandler extends WebCustomControlCoverageHandler_1.WebCustomControlCoverageHandler {
    constructor() {
        super(...arguments);
        this.type = "Form";
        // /**
        //  * Create a selector from this coverage handler to childDomNode
        //  * @param childDomNode
        //  * @param useExistingSelector
        //  * @returns a selector
        //  */
        // public createSelectorToChild(childDomNode : HTMLElement, useExistingSelector : boolean) : string {
        //     // Get our JQuery path from this element to the given child node
        //     let partialSelector : string = this.createSelectorFromParentToChild(this.domNode, childDomNode);
        //     // As we will find ourselves, this partial selector is enough
        //     return partialSelector;
        // }
    }
    startRecording() {
        throw new Error("Method not implemented.");
    }
    isRecordable() {
        return false;
    }
    isCapturable() {
        return false;
    }
    isListenable() {
        return false;
    }
    areChildrenCapturable() {
        return true;
    }
    areChildrenRecordable() {
        return true;
    }
}
exports.default = FormWebControlCoverageHandler;
