// @jacada-jbot/web/include/CoverageHandlers/Web/ImageWebControlCoverageHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const WebControlCoverageHandler_1 = require("./WebControlCoverageHandler");
class ImageWebControlCoverageHandler extends WebControlCoverageHandler_1.default {
    type() {
        return "WebImageControl";
    }
    executeStartRecordingOnNode() {
        // Do not record 
        return true;
    }
    /**
     * This method executes capture on this element.
     * It registers it for the highlight will recognize it, and sets the callback function that will be called
     * when the element will be clicked during capture.
     * @returns true
     */
    executeStartCapturingOnNode() {
        this.registerControlForHighlight(this, this.$currentNode, (event) => {
            this.fetchLabelIfExists();
            this.executeStartCapturingClicked(event);
        });
        // Do we have label from parent?
        this.fetchLabelIfExists();
        return true;
    }
    /**
     * This method creates a selector from this coverage handler to childDomNode
     * @param childDomNode
     * @param useExistingSelector
     * @returns a selector
     */
    createSelectorToChild(childDomNode, useExistingSelector) {
        // Get our JQuery path from this element to the given child node
        let partialSelector = this.createSelectorFromParentToChild(this.domNode, childDomNode);
        // As we will find ourselves, this partial selector is enough
        return partialSelector;
    }
    createElementRepositoryNode() {
        let elementNode = super.createElementRepositoryNode();
        elementNode.config.nodeName = this.domNode.nodeName;
        return elementNode;
    }
}
exports.default = ImageWebControlCoverageHandler;
