// @jacada-jbot/web/include/CoverageHandlers/Web/ListWebControlCoverageHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const WebControlCoverageHandler_1 = require("./WebControlCoverageHandler");
class ListWebControlCoverageHandler extends WebControlCoverageHandler_1.default {
    type() {
        return "WebListControl";
    }
    isNodeToBeAddedToRepository() {
        return false;
    }
    executeStartRecordingOnNode() {
        // Do not record 
        return true;
    }
    /**
     * This method creates a selector from this coverage handler to childDomNode
     * @param childDomNode
     * @param useExistingSelector
     * @returns a selector
     */
    createSelectorToChild(childDomNode, useExistingSelector) {
        // Get our JQuery path from this element to the given child node
        let partialSelector = this.createSelectorFromParentToChild(this.domNode, childDomNode);
        // As we will find ourselves, this partial selector is enough
        return partialSelector;
    }
    createElementRepositoryNode() {
        let elementNode = super.createElementRepositoryNode();
        elementNode.config.nodeName = this.domNode.nodeName;
        if (this.domNode.nodeName.toLowerCase() === 'ul' || this.domNode.nodeName.toLowerCase() === 'ol') {
            elementNode.isSelectable = false;
        }
        else {
            elementNode.isSelectable = true;
        }
        return elementNode;
    }
    getAutoGeneratedIdPrefix() {
        return "List";
    }
}
exports.default = ListWebControlCoverageHandler;
