// @jacada-jbot/web/include/CoverageHandlers/Web/TextAreaWebControlCoverageHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const EventDetails_1 = require("@jacada-jbot/base/include/Model/EventDetails");
const GenericResult_1 = require("@jacada-jbot/base/include/Model/GenericResult");
const LogicalAction_1 = require("@jacada-jbot/base/include/Model/LogicalAction");
const Log4jConfig_1 = require("@jacada-jbot/base/include/Utils/Log4jConfig");
const WebCustomControlCoverageHandler_1 = require("./WebCustomControlCoverageHandler");
const log = Log4jConfig_1.logFactory.getLogger("TextAreaWebControlCoverageHandler");
/**
 * Coverage handler for a textarea
 */
class TextAreaWebControlCoverageHandler extends WebCustomControlCoverageHandler_1.WebCustomControlCoverageHandler {
    constructor() {
        super(...arguments);
        this.type = "TextArea";
    }
    startRecording() {
        this.recordEvent("change", (event) => {
            let lAction = new LogicalAction_1.default(this.setText.name, event.target.value);
            return lAction;
        });
        return true;
    }
    subscribeToTextChanged() {
        this.listenToEvent("change", (event) => {
            return new EventDetails_1.default("textChanged", event.target.value);
        });
    }
    /**
     * This method gets the value of the node.
     * @returns a GenericResult struct
     */
    getText() {
        let text = this.node.val();
        return new GenericResult_1.GenericResult(true, null, text);
    }
    /**
     * This method sets the value with the given text by dispatching change event on the node.
     * @param text - the text to set
     * @returns a GenericResult struct
     */
    setText(text) {
        this.node.val(text);
        this.node[0].dispatchEvent(new Event("change"));
        return new GenericResult_1.GenericResult(true, null, text);
    }
}
exports.default = TextAreaWebControlCoverageHandler;
