// @jacada-jbot/web/include/WebCoverageHandlerFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("core-js/es/array");
// Collection of functions which are kept as their name in order to provide the Coverage Handler API to be called from JBot JS Engine.
function coverageHandlerFactoryInit(logger) {
    return true;
}
exports.coverageHandlerFactoryInit = coverageHandlerFactoryInit;
function createPlatformDriver(platformDriverObj) {
    return true;
}
exports.createPlatformDriver = createPlatformDriver;
function createRepositoryHandler(repositoryTreeJSON) {
    return true;
}
exports.createRepositoryHandler = createRepositoryHandler;
function createApplicationData(desktopData, sessionData, screenData, platformApplicationData) {
    return true;
}
exports.createApplicationData = createApplicationData;
function createApplicationCoverageHandler() {
    return true;
}
exports.createApplicationCoverageHandler = createApplicationCoverageHandler;
function createAutomationElement(winAe) {
    return true;
}
exports.createAutomationElement = createAutomationElement;
function startCapturing() {
}
exports.startCapturing = startCapturing;
